### ISSUE TITLE
Export Envoy version as string

### ISSUE THREAD

**Issue by dsymonds** [envoyproxy/envoy#13500]
*Title*: Export Envoy version as string

*Description*:
The `/stats` handler exports `server.version` (and `/stats/prometheus` likewise `envoy_server_version`), but it's a bare number (e.g. `9417675`), with no structure and is not human readable. In contrast, /server_info has a nicely structured `user_agent_build_version` block, but that isn't easily able to be scraped with Prometheus. The plain number is sufficient for spotting outliers (e.g. if not all Envoy instances in a fleet have been updated), but isn't pleasant to read.

It'd be nice if Envoy exported a stat that is a human-readable string (e.g. "1.16.0") on `/stats` and `/stats/prometheus`. This would make it easier to identify the state of outliers in a fleet, and also easier to analyse relative behaviour or performance when rolling out a new Envoy release.

**Comment by dio** [comment]
@dsymonds what do you think of overriding that value on your own (make sure it is integer)? e.g. set it as 1160 via [stats_server_version_override](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto#envoy-v3-api-field-config-bootstrap-v3-bootstrap-stats-server-version-override)

**Comment by dsymonds** [comment]
That's perhaps a partial solution, but it relies on active config changes rather than the binary identity merely being exposed. It'd be too easy to accidentally have the config set `server.version` to the wrong value, especially as configs otherwise generally work across Envoy versions.

**Comment by dio** [comment]
@dsymonds yeah, that's true. However, it depends on how you prepare the "bootstrap" config though. I think it is common to introduce a script to prepare the (bootstrap) config, hence we can introduce a function to avoid accidental setting `stats_server_version_override` with a wrong value. 

We can extract the version from the envoy binary, something like:

```
$ echo "stats_server_version_override: $(envoy --version | cut -d ':' -f 2 | cut -d '/' -f 2 | tr -d '\n' | tr -d '.' | cut -d '-' -f 1)" >> envoy-bootstrap-config.yaml
1160
```

(sorry my bash skill is bad).

**Comment by dsymonds** [comment]
I understand, though we don't operate that way. We already generate quite a large bootstrap config and bake it into Docker images. I guess we could try rigging up something to extract the version and mutate the config YAML file at container startup time, but that gets pretty messy.

(FWIW, `echo $(envoy --version | cut -d/ -f2)` is a lot shorter)

**Comment by mattklein123** [comment]
Don't we have a string stat type now? Can we add server version as a string stat type and then also fix Prometheus to output that type if not done already (IIRC that was not implemented but should be simple). cc @jmarantz

**Comment by jmarantz** [comment]
Yes, there's a TextReadout stat type, which accepts arbitrary strings.

**Comment by moderation** [comment]
Project Contour which is an Envoy based ingress controller for Kubernetes added this functionality a while back. The followed the example of Prometheus is terms of exposing the version - https://github.com/projectcontour/contour/issues/2378#issuecomment-602956666

**Comment by dsymonds** [comment]
no stalebot

**Comment by dio** [comment]
For `/stats`, adding an additional version as text readout should not be a problem. However, for Prometheus as mentioned in https://github.com/envoyproxy/envoy/issues/13500#issuecomment-707214699. We need to expose the version as a gauge with version text as a label of that gauge,  e.g. 

```
envoy_build_info{version: "1.7.0", revision: "bed0262ae96a40040dfd7403b5ad6392c3bfa20c"}
```

A quick solution for this probably handle text readout in `PrometheusStatsFormatter::statsAsPrometheus` and treat each entry of text readout metrics as a gauge where the value of it is the label value:

```
envoy_version_info{text_value: "1.7.0"}
```

**Comment by cgkades** [comment]
Has the work from that merge been deleted, or did it never solve this issue? Or do I just not understand how to use it?

```
~$ curl 'http://localhost:9901/stats?format=prometheus&text_readouts' | grep 'version'
<---snip---->
# TYPE envoy_server_version gauge
envoy_server_version{} 11405305
<---snip---->
```
